includes/modules/management/add/class-wcs-att-manage-add-product.php

```diff
@@ -0,0 +1,437 @@
+
+		if ( ! WCS_ATT_Product::supports_feature( $product, 'subscription_management_add_to_subscription' ) ) {
+			return;
+		}
+
+		// Bypass when switching.
+		if ( WCS_ATT_Manage_Switch::switching_product( $product ) ) {
+			return;
+		}
+
+		// User must be logged in.
+		if ( ! is_user_logged_in() ) {
+			return;
+		}
+
+		$subscription_options_visible = true;
+
+		/*
+		 * Subscription options for variable products are embedded inside the variation data 'price_html' field and updated by the core variations script.
+		 * The add-to-subscription template is displayed when a variation is found.
+		 */
+		if ( $product->is_type( 'variable' ) ) {
+
+			$subscription_options_visible = false;
+
+		} elseif ( WCS_ATT_Product::supports_feature( $product, 'subscription_management_add_to_matching_subscription' ) ) {
+
+			$subscription_options_visible = false;
+
+			$subscription_schemes                 = WCS_ATT_Product_Schemes::get_subscription_schemes( $product );
+			$force_subscription                   = WCS_ATT_Product_Schemes::has_forced_subscription_scheme( $product );
+			$is_single_scheme_forced_subscription = $force_subscription && sizeof( $subscription_schemes ) === 1;
+			$default_subscription_scheme_key      = apply_filters( 'wcsatt_get_default_subscription_scheme_id', WCS_ATT_Product_Schemes::get_default_subscription_scheme( $product, 'key' ), $subscription_schemes, false === $force_subscription, $product ); // Why 'false === $force_subscription'? The answer is back-compat.
+			$default_subscription_scheme          = WCS_ATT_Product_Schemes::get_subscription_scheme( $product, 'object', $default_subscription_scheme_key );
+
+			$subscription_options_visible = $is_single_scheme_forced_subscription || ( is_object( $default_subscription_scheme ) && ! $default_subscription_scheme->is_prorated() );
+		}
+
+		wc_get_template( 'single-product/product-add-to-subscription.php', array(
+			'product_id' => $product->get_id(),
+			'is_visible' => $subscription_options_visible
+		), false, WCS_ATT()->plugin_path() . '/templates/' );
+	}
+
+	/**
+	 * 'Add to subscription' view -- matching list of subscriptions.
+	 *
+	 * @param  array                $subscriptions
+	 * @param  WC_Product           $product
+	 * @param  WCS_ATT_Scheme|null  $scheme
+	 * @return void
+	 */
+	public static function matching_subscriptions_template( $subscriptions, $product, $scheme ) {
+
+		add_action( 'woocommerce_my_subscriptions_actions', array( __CLASS__, 'button_template' ) );
+
+		wp_nonce_field( 'wcsatt_add_product_to_subscription', 'wcsatt_nonce_' . $product->get_id() );
+
+		wc_get_template( 'single-product/product-add-to-subscription-list.php', array(
+			'subscriptions' => $subscriptions,
+			'product'       => $product,
+			'scheme'        => $scheme,
+			'user_id'       => get_current_user_id()
+		), false, WCS_ATT()->plugin_path() . '/templates/' );
+
+		remove_action( 'woocommerce_my_subscriptions_actions', array( __CLASS__, 'button_template' ) );
```

Comment:
Sometimes I think I prefer homogeneity than DRYness. A parent class would require a lot of context-related conditional code to litter these otherwise very simple templating functions, making the code less readable and harder to maintain (for me at least). 

Also both of these inherit from the `WCS_ATT_Abstract_Module` class, and I'd rather avoid a 3-level deep inheritance scheme for something that's basically just a few hooks organized in modules precisely in order to maintain separation (for readability and easier maintenance).